### PR TITLE
Update TypeScript and ESLint to latest versions

### LINKED ISSUE #818: Typescript version is outdated
## Requirement - what kind of business use case are you trying to solve?

We'd like to upgrade the typescript version from 3.5.3 to the latest one (4.4)

## Problem - what in Jaeger blocks you from solving the requirement?

the current version is missing key features

Related: #998

### PR DESCRIPTION
**Pull request by mszabo-wikia**
## Which problem is this PR solving?
- Resolves #818

## Short description of the changes
Update TypeScript, ESLint and `@typescript-eslint` to the latest versions.

Notable changes:
* Update the list of core ESLint rules that need to be disabled because `@typescript-eslint` provides TypeScript-specific equivalents for them, and change existing suppressions referencing the now-disabled core rules. Group these together in the ESLint config for clarity.
* Fix references to unimported types in `serviceGraph.tsx` and `types.tsx`.
* It seems that the type of `defaultConfig` and the user-defined config wasn't correctly inferred as `Config` before the upgrade. Fix various type errors resulting from this:
  * `'__mergeFields'`, i.e. the list of config options where user-specified values should be merged with the default object values in `defaultConfig`, was defined via `defineProperty` and so seemingly inferred as an unknown type. This now causes a type error when iterating over its keys to try and merge the relevant property values, as TypeScript can't guarantee that the property values are objects. Make it a separate named export instead, which also allows for strictly typing the allowable property names.
  * Make the `url` field of the `ConfigMenuItem` type optional as this was now recognized as a type error. This matches documentation and actual usage.
  * Rename the `dagMaxServicesLen` property definition to `dagMaxNumServices` as that is what the actual code uses.
  * Fix the `linkPatterns` config option typedef to be an array, as per actual usage.
 * Update `handleError` in Plexus to correctly take either an `ErrorEvent` or a `MessageEvent` to match updated typings for the Web Worker API.
 * Update `catch` clauses in `readJsonFile` because TypeScript 4.4+ [types the error in `catch` clauses as `unknown` by default](https://devblogs.microsoft.com/typescript/announcing-typescript-4-4/#using-unknown-in-catch-variables). We don't expect anything outside of an `Error` to be thrown here, so we can restore the previous behavior.

### CODE REVIEW

File: .eslintrc.js
Comment (yurishkuro) [inline .eslintrc.js:60]: Just curious, what's the harm in keeping them? if TS flags them first, then fixing the issue would also fix ESLint issue, would it not? (basically, I prefer fewer deviations from defaults if possible)

File: packages/jaeger-ui/src/components/DeepDependencies/Graph/DdgNodeContent/index.tsx
Comment (yurishkuro) [inline packages/jaeger-ui/src/components/DeepDependencies/Graph/DdgNodeContent/index.tsx:114]: it's good to have a comment explaining the reason (I assume it's a false positive)

File: packages/jaeger-ui/src/types/config.tsx
Comment (yurishkuro) [inline packages/jaeger-ui/src/types/config.tsx:95]: curious why these specific entries need `readonly` and not all others.

File: .eslintrc.js
Comment (mszabo-wikia) [reply .eslintrc.js:60]: The problem is that these can cause false positives with the new version of `@typescript-eslint`. E.g. removing the suppression for core's `no-shadow` rules results in a plethora of bogus reports about TS type definitions. The recommendation of silencing the corresponding core rules comes from the [plugin itself](https://typescript-eslint.io/rules/no-shadow/).

Comment (mszabo-wikia) [reply .eslintrc.js:60]: Another way to fix it could be to make the root eslintrc also extend the `recommended` config from `@typescript-eslint`, because that config already does this toggling for us. This is what plexus does already. I'm thinking this could be done in a followup PR as it might cause some new errors.

**Review by yurishkuro — APPROVED** [review]

**Comment by yurishkuro** [comment]
@mszabo-wikia can we also close #998 ?

File: packages/jaeger-ui/src/types/config.tsx
Comment (mszabo-wikia) [reply packages/jaeger-ui/src/types/config.tsx:95]: It seems it's specifically for `array` type entries. TS is apparently now better at detecting that the object passed to `deepFreeze` is recursively `readonly`.

File: .eslintrc.js
Comment (yurishkuro) [reply .eslintrc.js:60]: +1 -- #1258

**Comment by yurishkuro** [comment]
🥳 🥳 🥳 🥳 🥳 🥳 🥳 🥳 🥳 🥳 🥳 🥳 🥳 🥳 🥳 🥳 🥳 🥳

**Comment by yurishkuro** [comment]
Another awesome PR! It looks like it's going to unblock a lot of pending dependabot upgrades, several of them already turned green after failing CI for months. I'll be merging them today (unfortunately, every time I merge one I need to wait for the CI to run on all others, takes a lot of time).

**Comment by mszabo-wikia** [comment]
Awesome, that's great to see :)

**Pull request merged**
